### PR TITLE
Adds JUnit Rule support

### PR DESCRIPTION
**Pull request by lordofthejars**

### CODE REVIEW

File: arquillian-ape-nosql/couchbase/src/main/java/org/arquillian/ape/nosql/couchbase/CouchbasePopulatorService.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 import static org.awaitility.Awaitility.await;
 
-class CouchbasePopulatorService implements NoSqlPopulatorService<Couchbase> {
+public class CouchbasePopulatorService implements NoSqlPopulatorService<Couchbase> {
```
Comment (bartoszmajsak) [inline arquillian-ape-nosql/couchbase/src/main/java/org/arquillian/ape/nosql/couchbase/CouchbasePopulatorService.java]: Maybe if we follow SPI approach proposed in the comment in the rule itself we won't have to make those classes public?

File: arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java [hunk 2]
```diff
@@ -109,6 +109,11 @@ public void clean(List<String> resources) {
     private Optional<IDatabaseConnection> lookupDataSourceConnection(String dataSourceName) {
 
         try {
+
+            if (contextInstance == null) {
+                // In case of running outside Arquillian Runner
```
Comment (bartoszmajsak) [inline arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java]: Hmmm this is getting a bit ugly :) Having classes based on ARQ services / concepts used outside of its lifecycle we will end up with a lot of "ignore if `null`" - which is not really the nicest thing to have.

File: arquillian-ape-junit-rule/src/main/java/org/arquillian/ape/junit/rule/Reflection.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            .filter(field -> field.isAnnotationPresent(annotation))
+            .findFirst();
+    }
+
+    public static final void instantiateServiceAndPopulatorAndInject(Object testInstance, Field field, Class<? extends PopulatorService> serviceClass, Class<? extends Populator> populatorClass)
+        throws IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {
+        final Object service = serviceClass.newInstance();
+
+        // Using getDeclaredConstructor(serviceClass) does not work. Services must contain only one constructor, if not even the Arquillian runner fails in case of APE
+        // so it is not a constraint in this case at all.
```
Comment (bartoszmajsak) [inline arquillian-ape-junit-rule/src/main/java/org/arquillian/ape/junit/rule/Reflection.java]: I'm also not really fond of having such a method in general purpose `Reflections` utility class - maybe we can think of sth better?

File: arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java (hunk 2 above)
Comment (bartoszmajsak) [inline arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java]: Long term - we have to address this concern

File: arquillian-ape-junit-rule/src/main/java/org/arquillian/ape/junit/rule/Reflection.java (hunk 3 above)
Comment (lordofthejars) [inline arquillian-ape-junit-rule/src/main/java/org/arquillian/ape/junit/rule/Reflection.java]: Well forme this class is where all Reflection code resides, not like a util class.

File: arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java (hunk 2 above)
Comment (lordofthejars) [inline arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java]: I agree but for this specific use case I think we can leave as is for now. In long term we could inject some `stubbed` classes.

Comment (bartoszmajsak) [inline arquillian-ape-sql/core-dbunit/src/main/java/org/arquillian/ape/rdbms/dbunit/DbUnitPopulatorService.java]: If only that will be the only use case ;) Let's see how it evolves

**Pull request merged**